[iOS 17] Build on the public SDK

emw-apple opened the pull request:
#### aa721f7998fcfe5bba32f0feb9b7217c7ba520da
<pre>
[iOS 17] Build on the public SDK
<a href="https://bugs.webkit.org/show_bug.cgi?id=259513">https://bugs.webkit.org/show_bug.cgi?id=259513</a>

Reviewed by Alexey Proskuryakov.

Adds iphoneos17.0-additions and iphonesimulator17.0-additions SDKs
containing stubs for private frameworks. Unlike past cycles, these have
been generated automatically from the internal SDK, using a `strip-tbd`
program (also added). The workflow I used was:

1. Using products from an internal build, `otool -L` relevant binaries
   (WebCore, WebKit, TestWebKitAPI) to find private frameworks linked
   against.

2. Copy those frameworks&apos; tbd stubs from the internal SDK to the same
   locations in the sparse SDK.

3. Strip out all the symbols we aren&apos;t binding to:

    nm -gj WebKitBuild/Debug/WebCore.framework/WebCore &gt; /tmp/WebKit.exp
    nm -gj WebKitBuild/Debug/WebKit.framework/WebKit &gt;&gt; /tmp/WebKit.exp
    # etc.
    strip-tbd -s /tmp/WebKit.exp WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/**/*.tbd

Additionally, this workflow has been automated into a script which generates the API stubs
for a sparse SDK in a single invocation. I generated iphonesimulator17.0
framework stubs by running:

    extract-tbds-from-internal-sdk -b $build `xcrun --show-sdk-path --sdk iphonesimulator.internal` WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk

The new linker included in Xcode 15 is stricter about validating targets
in TBDs (it uses LLVM&apos;s codepath for parsing TBDs, which treats unknown
platforms as an error). As a result,
configure-xcode-for-embedded-development needs to patch TBDs in the base
SDK with *target* strings, not just architectures. This also means that
the private framework stubs cannot be shared between device and
simulator variants, but since we can now auto-generate the stubs, I
think it&apos;s reasonable to check in separate TBDs for both SDKs.

Aside from the sparse SDK changes, make a few minor fixups.

* Source/WTF/wtf/PlatformHave.h: Lockdown mode was never supported in
  simulators, and a Core Graphics function we call in LDM was removed
  from the simulator SDK. Refine this platform macro.
* Source/WebKit/NetworkProcess/cocoa/AppStoreDaemonSPI.h: The public
  side was missing NONNULL markers.
* Tools/Scripts/configure-xcode-for-embedded-development: Patch TBDs
  with target strings, as mentioned above.
* Tools/Scripts/extract-tbds-from-internal-sdk: Added.
* Tools/Scripts/strip-tbd: Added.
* Tools/WebKitTestRunner/Configurations/Base.xcconfig:
  In non-macOS builds, /System/Library/PrivateFrameworks is being given
  as a framework search path, causing frameworks on the *host* machine
  to be preferred over system frameworks. Prefix it with $(SDK_DIR),
  which is probably what was originally intended.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/SDKSettings.plist: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/AppServerSupport.framework/AppServerSupport.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/AppStoreDaemon.framework/AppStoreDaemon.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/AppSupport.framework/AppSupport.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/ApplePushService.framework/ApplePushService.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/AuthKit.framework/AuthKit.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/BackBoardServices.framework/BackBoardServices.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/CorePrediction.framework/CorePrediction.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/FontServices.framework/FontServices.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/FontServices.framework/libFontParser.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/FrontBoardServices.framework/FrontBoardServices.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/InstallCoordination.framework/InstallCoordination.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/PrototypeTools.framework/PrototypeTools.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/RunningBoardServices.framework/RunningBoardServices.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/SafariSafeBrowsing.framework/SafariSafeBrowsing.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/URLFormatting.framework/URLFormatting.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/VisionKitCore.framework/VisionKitCore.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/usr/local/include/AvailabilityProhibitedInternal.h: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/SDKSettings.plist: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/AppServerSupport.framework/AppServerSupport.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/AppStoreDaemon.framework/AppStoreDaemon.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/AppSupport.framework/AppSupport.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/ApplePushService.framework/ApplePushService.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/AuthKit.framework/AuthKit.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/BackBoardServices.framework/BackBoardServices.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/CorePrediction.framework/CorePrediction.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/FontServices.framework/libFontParser.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/FrontBoardServices.framework/FrontBoardServices.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/IOSurfaceAccelerator.framework/IOSurfaceAccelerator.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/InstallCoordination.framework/InstallCoordination.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/PrototypeTools.framework/PrototypeTools.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/RunningBoardServices.framework/RunningBoardServices.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/SafariSafeBrowsing.framework/SafariSafeBrowsing.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/URLFormatting.framework/URLFormatting.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/VisionKitCore.framework/VisionKitCore.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/usr/local/include/AvailabilityProhibitedInternal.h: Copied from other SDK additions.

Canonical link: <a href="https://commits.webkit.org/266551@main">https://commits.webkit.org/266551@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/70578b384cf42e958f8937bb5c4ed8762e3248cc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14088 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14403 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14740 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15828 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13358 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14174 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16913 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14487 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/16033 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14260 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14843 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11946 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/16539 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12126 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12707 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19727 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/12035 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13205 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12871 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16074 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/13351 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13411 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11280 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/14132 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12690 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3657 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3415 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17023 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/14519 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13253 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3470 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

aproskuryakov reviewed (approved):
rs=me, although I'm a bit uncertain about the premise.

> This also means that
the private framework stubs cannot be shared between device and
simulator variants

Aren't stubs shared in the normal SDK? What is different about what we do that makes it impossible? I'm concerned that people will find it challenging to add new symbols to all relevant stubs.

emw-apple committed:
[iOS 17] Build on the public SDK
https://bugs.webkit.org/show_bug.cgi?id=259513

Reviewed by Alexey Proskuryakov.

Adds iphoneos17.0-additions and iphonesimulator17.0-additions SDKs
containing stubs for private frameworks. Unlike past cycles, these have
been generated automatically from the internal SDK, using a `strip-tbd`
program (also added). The workflow I used was:

1. Using products from an internal build, `otool -L` relevant binaries
   (WebCore, WebKit, TestWebKitAPI) to find private frameworks linked
   against.

2. Copy those frameworks' tbd stubs from the internal SDK to the same
   locations in the sparse SDK.

3. Strip out all the symbols we aren't binding to:

    nm -gj WebKitBuild/Debug/WebCore.framework/WebCore > /tmp/WebKit.exp
    nm -gj WebKitBuild/Debug/WebKit.framework/WebKit >> /tmp/WebKit.exp
    # etc.
    strip-tbd -s /tmp/WebKit.exp WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/**/*.tbd

Additionally, this workflow has been automated into a script which generates the API stubs
for a sparse SDK in a single invocation. I generated iphonesimulator17.0
framework stubs by running:

    extract-tbds-from-internal-sdk -b $build `xcrun --show-sdk-path --sdk iphonesimulator.internal` WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk

The new linker included in Xcode 15 is stricter about validating targets
in TBDs (it uses LLVM's codepath for parsing TBDs, which treats unknown
platforms as an error). As a result,
configure-xcode-for-embedded-development needs to patch TBDs in the base
SDK with *target* strings, not just architectures. This also means that
the private framework stubs cannot be shared between device and
simulator variants, but since we can now auto-generate the stubs, I
think it's reasonable to check in separate TBDs for both SDKs.

Aside from the sparse SDK changes, make a few minor fixups.

* Source/WTF/wtf/PlatformHave.h: Lockdown mode was never supported in
  simulators, and a Core Graphics function we call in LDM was removed
  from the simulator SDK. Refine this platform macro.
* Source/WebKit/NetworkProcess/cocoa/AppStoreDaemonSPI.h: The public
  side was missing NONNULL markers.
* Tools/Scripts/configure-xcode-for-embedded-development: Patch TBDs
  with target strings, as mentioned above.
* Tools/Scripts/extract-tbds-from-internal-sdk: Added.
* Tools/Scripts/strip-tbd: Added.
* Tools/WebKitTestRunner/Configurations/Base.xcconfig:
  In non-macOS builds, /System/Library/PrivateFrameworks is being given
  as a framework search path, causing frameworks on the *host* machine
  to be preferred over system frameworks. Prefix it with $(SDK_DIR),
  which is probably what was originally intended.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/SDKSettings.plist: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/AppServerSupport.framework/AppServerSupport.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/AppStoreDaemon.framework/AppStoreDaemon.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/AppSupport.framework/AppSupport.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/ApplePushService.framework/ApplePushService.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/AuthKit.framework/AuthKit.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/BackBoardServices.framework/BackBoardServices.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/CorePrediction.framework/CorePrediction.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/FontServices.framework/FontServices.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/FontServices.framework/libFontParser.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/FrontBoardServices.framework/FrontBoardServices.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/InstallCoordination.framework/InstallCoordination.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/PrototypeTools.framework/PrototypeTools.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/RunningBoardServices.framework/RunningBoardServices.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/SafariSafeBrowsing.framework/SafariSafeBrowsing.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/URLFormatting.framework/URLFormatting.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/VisionKitCore.framework/VisionKitCore.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/usr/local/include/AvailabilityProhibitedInternal.h: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/SDKSettings.plist: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/AppServerSupport.framework/AppServerSupport.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/AppStoreDaemon.framework/AppStoreDaemon.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/AppSupport.framework/AppSupport.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/ApplePushService.framework/ApplePushService.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/AuthKit.framework/AuthKit.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/BackBoardServices.framework/BackBoardServices.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/CorePrediction.framework/CorePrediction.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/FontServices.framework/libFontParser.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/FrontBoardServices.framework/FrontBoardServices.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/IOSurfaceAccelerator.framework/IOSurfaceAccelerator.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/InstallCoordination.framework/InstallCoordination.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/PrototypeTools.framework/PrototypeTools.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/RunningBoardServices.framework/RunningBoardServices.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/SafariSafeBrowsing.framework/SafariSafeBrowsing.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/URLFormatting.framework/URLFormatting.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/VisionKitCore.framework/VisionKitCore.tbd: Added.
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/usr/local/include/AvailabilityProhibitedInternal.h: Copied from other SDK additions.

Canonical link: https://commits.webkit.org/266551@main
